### PR TITLE
fix: add field descriptor to provision failed log statement

### PR DESCRIPTION
**Pull request by ben-fussell**
## Why?
- Field provisioning errors don't explain which field failed

## What changed?
- Add field descriptor to error message

### CODE REVIEW

File: contxt/cli/commands/iot.py [hunk 1]
```diff
@@ -198,7 +198,11 @@ def create(clients: Clients, feed_key: str, input: IO[str]) -> None:
             field = cast(Field, field)
             if field.field_descriptor not in curr_fields:
                 # New field, create it
-                fields[i][0] = clients.iot.provision_field_for_feed(feed.id, field)
+                try:
+                    fields[i][0] = clients.iot.provision_field_for_feed(feed.id, field)
+                except HTTPError as e:
+                    logging.error(f"Error provisioning field: {field.field_descriptor}")
+                    raise e
```
Comment (andreimoustache) [inline contxt/cli/commands/iot.py:205]: Could also accumulate the failed ones in a list and then do something with them later. Like print the list of failed, and exit with a non-zero status code.

```suggestion
                    failed_requests.append(e)
```

**Review by andreimoustache — APPROVED** [review]

**Comment by ben-fussell** [comment]
@andreimoustache I added the failure tracking concept like you mentioned. Now, for each step in the provisioning process errors are logged after the entire list is processed.

I also made each step dependent on success of the previous. So in the process below, subsequent steps don't execute until the previous step is error free.

`Provision fields -> add fields to groupings -> add groupings to categories`

Let me know if you agree with all this or have any other suggestions before I merge. Thanks!

File: contxt/cli/commands/iot.py [hunk 2]
```diff
@@ -238,6 +282,25 @@ def create(clients: Clients, feed_key: str, input: IO[str]) -> None:
...
                 logging.debug(e)
+                failures.append(
+                    {
+                        "grouping_id": grouping_id,
+                        "category_id": category_id,
+                        "error_message": e.response.text,
+                    }
+                )
+    if len(failures) > 0:
+        printProvisioningFailures(
```
Comment (andreimoustache) [inline contxt/cli/commands/iot.py:293]:
```suggestion
        print_provisioning_failures(
```

File: contxt/cli/commands/iot.py [hunk 3]
```diff
@@ -226,10 +241,39 @@ def create(clients: Clients, feed_key: str, input: IO[str]) -> None:
...
 
     # Add groupings to categories
     categories = {c.name: c for c in clients.iot.get_categories_for_facility(feed.facility_id)}
     new_groups = {g: c for f, g, c in fields}
+
+    # Verify that all categories exist
+    new_categories = list(set(v for v in new_groups.values()))
+    missing_categories = [{"category": c} for c in new_categories if c not in categories]
+    if len(missing_categories) > 0:
+        printProvisioningFailures(
```
Comment (andreimoustache) [inline contxt/cli/commands/iot.py:269]:
```suggestion
        print_provisioning_failures(
```

**Comment by andreimoustache** [comment]
Hey, awesome!

> I also made each step dependent on success of the previous. So in the process below, subsequent steps don't execute until the previous step is error free.
> 
> `Provision fields -> add fields to groupings -> add groupings to categories`

Sounds great, I think!

Only nit, and it's minor, would be to `raise` instead of `return`. Not 100% what the nicest way to go would be with Click, but I see we have some `raise click.ClickException`s.

**Comment by ben-fussell** [comment]
Thanks @andreimoustache! I'm stuck on javascript naming lol...

I'll test with `raise click.ClickException` to confirm and then merge

**Comment by andreimoustache** [comment]
Nah, you're all right. We just need to put the linter to work

**Review by andreimoustache — APPROVED** [review]
https://gfycat.com/personalzanyfalcon

**Pull request merged**